packages/daemon/src/service/trace-manager.ts

```diff
@@ -0,0 +1,208 @@
+import { Transform, TransformCallback } from 'stream';
+import { EventEmitter } from 'events';
+import { open, close, write } from 'fs-extra';
+import { provide, scope, ScopeEnum } from 'midway';
+import { StringDecoder } from 'string_decoder';
+import { generateId, PipelineStatus, PluginTypeI } from '@pipcook/pipcook-core';
+
+export type PipcookEventType = 'log' | 'job_status';
+
+export interface JobStatusChangeEvent {
+  jobStatus: PipelineStatus;
+  step?: PluginTypeI;
+  stepAction?: 'start' | 'end';
+}
+
+export interface LogEvent {
+  level: string;
+  data: string;
+}
+
+export type PipecookEvent = JobStatusChangeEvent | LogEvent;
+/**
+ * tracer for plugin installing and pipeline running.
+ */
+export class Tracer {
+  // log id
+  id: string;
+  // stdout stream for log pipe
+  private stdout: LogPassthrough;
+  // stderr stream for log pipe
+  private stderr: LogPassthrough;
+  // event emitter
+  private emitter: EventEmitter;
+
+  private opts?: LogOptions;
+  private fdOut: number;
+  private fdErr: number;
+  constructor(opts?: LogOptions) {
+    this.id = generateId();
+    this.emitter = new EventEmitter();
+    this.opts = opts;
+  }
+
+  /**
+   * init loggers
+   */
+  async initLogger() {
+    this.fdOut = this.opts?.stdoutFile ? await open(this.opts?.stdoutFile, 'w+') : -1;
+    this.fdErr = this.opts?.stderrFile ? await open(this.opts?.stderrFile, 'w+') : -1;
+    this.stdout = new LogPassthrough(this.fdOut);
+    this.stderr = new LogPassthrough(this.fdErr);
+  }
+  /**
+   * get the loggers
+   */
+  getLogger(): { stdout: LogPassthrough; stderr: LogPassthrough } {
+    return { stdout: this.stdout, stderr: this.stderr };
+  }
+
+  /**
+   * listen event
```

Comment:
Also, explain what's the event in a trace object.

Reply:
Updated, put the annotation on the class.